pkg/integrations/v4/dm/id_provider.go

```diff
@@ -8,98 +8,127 @@ import (
-	unregisteredEntities = make(UnregisteredEntities, 0)
-	registeredEntities = make(RegisteredEntitiesNameToID, 0)
+func (p *cachedIdProvider) ResolveEntities(entities []protocol.Entity) (registeredEntitiesNameToID, unregisteredEntityList) {
+	unregisteredEntities := make(unregisteredEntityList, 0)
+	registeredEntities := make(registeredEntitiesNameToID, 0)
 	entitiesToRegister := make([]protocol.Entity, 0)
 
+	// add error cache checking
+
 	for _, e := range entities {
+		p.cacheMutex.Lock()
 		if foundID, ok := p.cache[e.Name]; ok {
 			registeredEntities[e.Name] = foundID
+		} else if foundUnregisteredEntities, ok := p.unregisteredEntities[e.Name]; ok {
+			unregisteredEntities = append(unregisteredEntities, foundUnregisteredEntities)
 		} else {
+			unregisteredEntities = append(unregisteredEntities, newUnregisteredEntity(e, reasonNotInCache, nil))
 			entitiesToRegister = append(entitiesToRegister, e)
 		}
+		p.cacheMutex.Unlock()
 	}
 
-	if len(entitiesToRegister) == 0 {
-		return registeredEntities, unregisteredEntities
+	if len(entitiesToRegister) != 0 {
+		p.toRegisterQueue <- entitiesToRegister
 	}
 
-	response, _, errClient := p.client.RegisterBatchEntities(
-		agentIdn.ID,
-		entitiesToRegister)
+	return registeredEntities, unregisteredEntities
+}
 
+// todo add close chan logic
+func (p *cachedIdProvider) registerEntitiesWorker() {
 	type nameToEntityType map[string]protocol.Entity
-	nameToEntity := make(nameToEntityType, len(entitiesToRegister))
 
-	for i := range entitiesToRegister {
-		nameToEntity[entitiesToRegister[i].Name] = entitiesToRegister[i]
-	}
+	for {
+		select {
+		case entitiesToRegister := <-p.toRegisterQueue:
+			response, _, errClient := p.client.RegisterBatchEntities(
+				p.agentIdentity().ID,
+				entitiesToRegister)
 
-	if errClient != nil {
-		for i := range entitiesToRegister {
-			unregisteredEntity := newUnregisteredEntity(entitiesToRegister[i], reasonClientError, errClient)
-			p.unregisteredEntities[entitiesToRegister[i].Name] = unregisteredEntity
-			unregisteredEntities = append(unregisteredEntities, unregisteredEntity)
-		}
-	} else {
-		for i := range response {
-			if response[i].Err != "" {
-				unregisteredEntity := newUnregisteredEntity(nameToEntity[response[i].Name], reasonEntityError, fmt.Errorf(response[i].Err))
-				p.unregisteredEntities[response[i].Name] = unregisteredEntity
-				unregisteredEntities = append(unregisteredEntities, unregisteredEntity)
-				continue
+			nameToEntity := make(nameToEntityType, len(entitiesToRegister))
+			for i := range entitiesToRegister {
+				nameToEntity[entitiesToRegister[i].Name] = entitiesToRegister[i]
 			}
 
-			p.cache[string(response[i].Key)] = response[i].ID
-			registeredEntities[string(response[i].Key)] = response[i].ID
+			if errClient != nil {
+				for i := range entitiesToRegister {
+					p.cacheMutex.Lock()
```

Comment:
move the lock to outside the for loop. Would be easier to lock and update it all in one go than constantly trying to get a lock.